org.adempiere.base/src/org/compiere/model/MRole.java

```diff
@@ -2897,9 +2897,9 @@ private void loadChildRoles(boolean reload)
 	 */
 	private void loadSubstitutedRoles(boolean reload)
 	{
-		if (this.m_parent != null)
+		if (this.m_parent != null || isMasterRole())
```

Comment:
isn't the error is cause by user and substitute user is using the same master role instead of the fact the user is using a master role ?

Reply:
My finding was that loadSubstitutedRoles is being called multiple times, and it becomes recursive.

loadSubstitutedRoles just need to be called one time, because is user dependent, not role dependent, so the loadSubstitutedRoles must work just on the login role, not on children (which was already excluded) or master roles (which was not)

BTW, I think probably we have discussed this before, but the design of substitute roles seems wrong, however, is out of the scope of the ticket, would be worthy to discuss it in another ticket?

Reply:
hmm .. I think the fix should be along the line where if the substitute role = current role or have been loaded as child/included role, we should not try to include it again.

Reply:
I tried that as well, changing the whereClause in loadSubstitutedRoles, with the following patch:
[badTry.txt](https://github.com/idempiere/idempiere/files/9324393/badTry.txt)
But, it didn't work.

Also, there is where I found that the substitute roles is somewhat wrongly designed, because the substitute is related to the user, and not to the login role.
From what I see it is just wrong to add options from a role that is not selected on Login (which is the actual behavior of substitute roles).

I think what substitute role must do is to allow the user to login with the roles from the user being substituted.  With the actual behaviour, the substitute user is presented with a different menu than the substituted user when using exactly the same login role (I find that concerning).
The other point to use substituted roles is when doing a workflow approval.

But again, that's beyond the current ticket fix, the fix here is just to solve a bad recursion scenario.